symengine/polys/basic_conversions.h

```diff
@@ -229,6 +228,217 @@ RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex)
 }
+
+template <typename P>
+enable_if_t<std::is_same<MIntPoly, P>::value, typename P::container_type>
+_basic_to_mpoly(const RCP<const Basic> &basic, const set_basic &gens);
+
+template <typename P, typename V>
+class BasicToMPolyBase : public BaseVisitor<V>
+{
+public:
+    using Dict = typename P::container_type;
+    using Vec = typename Dict::vec_type;
+    Dict dict;
+    set_basic gens;
+    std::unordered_map<RCP<const Basic>, vec_basic, RCPBasicHash, RCPBasicKeyEq>
+        gens_pow;
+    umap_basic_uint gens_map;
+
+    Dict apply(const Basic &b, const set_basic &gens_)
+    {
+        gens = gens_;
+        dict.vec_size = gens.size();
+
+        RCP<const Basic> genpow, genbase;
+        unsigned int i = 0;
+
+        for (auto it : gens) {
+            genpow = one;
+            genbase = it;
+            if (is_a<const Pow>(*it)) {
+                genpow = rcp_static_cast<const Pow>(it)->get_exp();
+                genbase = rcp_static_cast<const Pow>(it)->get_base();
+            }
+            auto ite = gens_pow.find(genbase);
+            if (ite == gens_pow.end())
+                gens_pow[genbase] = {genpow};
+            else
+                gens_pow[genbase].push_back(genpow);
+            gens_map[it] = i++;
+        }
+
+        b.accept(*this);
+        return std::move(dict);
+    }
+
+    void dict_set(Vec pow, const Basic &x)
+    {
+        static_cast<V *>(this)->dict_set(pow, x);
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Integer>(*x.get_exp())) {
+            int i = rcp_static_cast<const Integer>(x.get_exp())->as_int();
+            if (i > 0) {
+                dict = Dict::pow(_basic_to_mpoly<P>(x.get_base(), gens), i);
+                return;
+            }
+        }
+
+        Vec zero_v(gens.size(), 0);
+        RCP<const Basic> coef = one, tmp;
+        RCP<const Integer> i;
+        bool found;
+        auto ite = gens_pow.find(x.get_base());
+
+        if (ite != gens_pow.end()) {
+
+            set_basic expos;
+
+            if (is_a<const Add>(*x.get_exp())) {
+                RCP<const Add> addx = rcp_static_cast<const Add>(x.get_exp());
+                for (auto const &it : addx->dict_)
+                    expos.insert(mul(it.first, it.second));
+                if (not addx->coef_->is_zero())
+                    expos.insert(addx->coef_);
+            } else {
+                expos.insert(x.get_exp());
+            }
+
+            for (auto const &it : expos) {
+
+                found = false;
+
+                for (auto powr : ite->second) {
+                    tmp = div(it, powr);
+                    if (is_a<const Integer>(*tmp)) {
+                        i = rcp_static_cast<const Integer>(tmp);
+                        if (i->is_positive()) {
+                            zero_v[gens_map[pow(ite->first, powr)]]
+                                = i->as_int();
+                            found = true;
```

Comment:
No, consider `2**(x + y)` where the only generator is `2**y`. At `x` the coeff will be updated to `2**x`, and at `y`, we don't want to make the coeff = 1. (It should throw, for MIntPolys)